### PR TITLE
Sentence sentiment analysis

### LINKED ISSUE #28: Sentiment analysis
Computing sentence based sentiment analysis and checking how sentiment changes over time for topics like heat pumps and boilers.

### LINKED ISSUE #32: Using an alternative sentiment model that either outputs the sentiment value or 3 categories: pos/neg/neutral
Current model only outputs pos/neg, but not neutral.

### PR DESCRIPTION
**Pull request by sofiapinto**
---

# Description

This PR was the initial PR created to add a pipeline for sentiment analysis (using Flair, which outputs positive/negative labels). We later edited it to use a different model - the Twitter-roBERTa-base for Sentiment Analysis.

Fixes #28 #32 

# Instructions for Reviewer

Hey @helloaidank , there are a couple of scripts in this PR but I'll ask you to review only a few of them:

- `asf_public_discourse_home_decarbonisation/pipeline/sentiment/README.md`
Can you please go through this README and let me know if anything important is missing?

- `asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py`
This is the script to compute sentiment using Twitter-roBERTa-base, which is the one we're currently using in our outputs (not Flair). This is an adapted version of code written by Liz, so no need to review that in detail as I have reviewed her code myself. Could you please read through the script and check if the documentation is enough?

If you could also try and run this to make sure it works on your end: 
`python asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py --source "mse" --filter_by_expression "heat pump" --start_date "2016-01-01" --end_date "2024-05-23" --test`

- `asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py`
Could you please check this script and documentation?
If you could also try and run this to make sure it works on your end: 
`python asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py --start_date "2018-01-01" --end_date "2024-05-23" --test`

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+```
+from asf_public_discourse_home_decarbonisation.pipeline.sentiment.sentence_sentiment import SentenceBasedSentiment
+sentiment_model = SentenceBasedSentiment()
+texts = ["This is a really great sentence", "This sentence is awful", "Cat"]
+sentiment_scores = sentiment_model.get_sentence_sentiment(texts)
+>> [("This is a really great sentence", 'positive', 0.97741115), ("This sentence is awful", 'negative', 0.9255473), ("Cat", 'neutral', 0.6470574)]
+```
+
+The output is a list of tuples where each tuple contains the sentence, sentiment label and the probability of the sentiment label.
```
Comment (helloaidank) [inline asf_public_discourse_home_decarbonisation/pipeline/sentiment/README.md]:
```suggestion
The output is a list of tuples where each tuple contains the sentence, sentiment label and the probability of the sentiment label. The higher the probability, the more certain we can be that the sentiment has been correctly classified. 
```

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+class SentenceBasedSentiment(object):
+    """Find sentiment scores for sentences"""
+
+    def __init__(
+        self,
+        model_name="cardiffnlp/twitter-roberta-base-sentiment-latest",
+    ):
+        """
+        Args:
+            model_name (str, optional): Model name. Defaults to "cardiffnlp/twitter-roberta-base-sentiment-latest".
```
Comment (helloaidank) [inline asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py:60]: Would probably good to know what class of sentiment models are compatible with this script?

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            model_name, use_fast=True, model_max_length=512
+        )
+        self.config = AutoConfig.from_pretrained(model_name)
+        self.model = AutoModelForSequenceClassification.from_pretrained(model_name)
+
+    def get_sentence_sentiment(self, texts: Union[list, str]) -> list:
+        """
+        Get the sentiment for a list of sentences using a pretrained sentiment model.
+        Returns a list of tuples with sentence, sentiment and the score:
+        e.g. ['sentence AAA', 'neutral', 0.6559918), ('sentence BBB', 'negative', 0.6319174)]
```
Comment (helloaidank) [inline asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py]:
```suggestion
        e.g. [("This is a really great sentence", 'positive', 0.97741115), ("This sentence is awful", 'negative', 0.9255473), ("Cat", 'neutral', 0.6470574)]
```
Think the example you used above is more informative, I'd find this "sentence AAA" slightly confusing.

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        type=str,
+    )
+    parser.add_argument(
+        "--test",
+        help="Run in test mode",
+        action="store_true",
+    )
+    return parser.parse_args()
+
+
```
Comment (helloaidank) [inline asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py:58]: Not sure if you had in mind for this code to be reviewed or refactored but just suggested refactoring here anyhow.
```
def preprocess_sentences(mse_data: pd.DataFrame, technology_name: str, args: argparse.Namespace) -> List[str]:
    # Preprocess data for the current technology
    sentences_data = prepping_data_for_topic_analysis(
        mse_data,
        technology_name,
        args.start_date,
        args.end_date,
        only_keep_sentences_with_expression=True,
    )

    # Limit data if in test mode
    if args.test:
        sentences_data = sentences_data.head(50)

    # Drop duplicates and extract sentences
    sentences = list(sentences_data.drop_duplicates("sentences")["sentences"])
    return sentences

def analyse_sentiment(
    sentences: List[str],
    sentiment_model: SentenceBasedSentiment,
    chunk_size: int = 100,
    technology_name: str = ""
) -> List[Tuple[str, str, float]]:
    # Analyse sentiment in chunks
    all_sentiment = []
    for text_chunk in tqdm(
        list_chunks(sentences, chunk_size=chunk_size),
        desc=f"Processing {technology_name}",
    ):
        sentiment_scores = sentiment_model.get_sentence_sentiment(text_chunk)
        all_sentiment.extend(sentiment_scores)
    return all_sentiment

def save_results(
    all_sentiment: List[Tuple[str, str, float]],
    path_to_save_prefix: str,
    output_suffix: str,
    args: argparse.Namespace
) -> pd.DataFrame:
    # Create DataFrame from results
    all_sentiment_df = pd.DataFrame(
        all_sentiment, columns=["text", "sentiment", "score"]
    )

    logging.info(f"Processed {len(all_sentiment_df)} sentences for {output_suffix}")
    logging.debug(all_sentiment_df.head())

    # Construct output file name
    output_name = (
        f"{path_to_save_prefix}_mse_{output_suffix}_sentences_sentiment.csv"
    )

    # Save results if not in test mode
    if not args.test:
        save_to_s3(
            S3_BUCKET,
            all_sentiment_df,
            output_name,
        )
    return all_sentiment_df

def process_technology(
    mse_data: pd.DataFrame,
    technology_name: str,
    output_suffix: str,
    sentiment_model: SentenceBasedSentiment,
    args: argparse.Namespace,
    path_to_save_prefix: str,
    chunk_size: int = 100,
) -> pd.DataFrame:
    try:
        # Preprocess sentences
        sentences = preprocess_sentences(mse_data, technology_name, args)

        # Analyse sentiment
        all_sentiment = analyse_sentiment(
            sentences, sentiment_model, chunk_size, technology_name
        )

        # Save results
        all_sentiment_df = save_results(
            all_sentiment, path_to_save_prefix, output_suffix, args
        )
        return all_sentiment_df
    except Exception as e:
        logging.error(f"Error processing {technology_name}: {e}")

```

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+if __name__ == "__main__":
+    chunk_size = 100
+    parser = ArgumentParser()
+    args = parse_arguments(parser)
+
+    mse_data = get_mse_data(
+        category="all", collection_date=config["latest_data_collection_date"]["mse"]
+    )
+
```
Comment (helloaidank) [inline asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py:63]:
```
    technologies = [
        {"name": "heat pump", "output_suffix": "heat_pump"},
        {"name": "solar panel", "output_suffix": "solar_panel"},
        {"name": "boiler", "output_suffix": "boiler"},
    ]

    for tech in technologies:
        process_technology(
            mse_data,
            tech["name"],
            tech["output_suffix"],
            sentiment_model,
            args,
            path_to_save_prefix,
            chunk_size=chunk_size,
        )
```

**Review by helloaidank — COMMENTED** [review]
Hi @sofiapinto, 
I have reviewed the following files:
- `asf_public_discourse_home_decarbonisation/pipeline/sentiment/README.md`
- `asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py`
- `asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py`
and ran over the python scripts without issue.
 I have left a few comments, with some suggestions for refactoring the `sentence_sentiment_technologies.py` script.  Otherwise all of it looks good.

Good work!

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py (hunk 3 above)
Comment (sofiapinto) [reply asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment.py]: good shout :)

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py (hunk 5 above)
Comment (sofiapinto) [reply asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py:63]: nice, will implement :)

File: asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py (hunk 4 above)
Comment (sofiapinto) [reply asf_public_discourse_home_decarbonisation/pipeline/sentiment/sentence_sentiment_technologies.py:58]: nice! will implement :)

**Pull request merged**